File: android/src/main/java/me/nathanfallet/suitebde/features/shop/ShopItemView.kt
package me.nathanfallet.suitebde.features.shop

import androidx.compose.foundation.ExperimentalFoundationApi
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ArrowBack
import androidx.compose.material3.*
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.livedata.observeAsState
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import me.nathanfallet.suitebde.R
import me.nathanfallet.suitebde.features.root.OldRootViewModel
import me.nathanfallet.suitebde.models.ensisa.TicketConfiguration
import me.nathanfallet.suitebde.ui.components.Picker

@OptIn(ExperimentalFoundationApi::class, ExperimentalMaterial3Api::class)
@Composable
fun ShopItemView(
    modifier: Modifier = Modifier,
    viewModel: ShopItemViewModel,
    oldRootViewModel: OldRootViewModel,
    navigateUp: () -> Unit,
) {

    val user by oldRootViewModel.getUser().observeAsState()

    val payNow by viewModel.getPayNow().observeAsState()
    val loading by viewModel.getLoading().observeAsState()
    val error by viewModel.getError().observeAsState()
    val success by viewModel.getSuccess().observeAsState()

    LazyColumn(modifier) {
        stickyHeader {
            TopAppBar(
                title = { Text(text = viewModel.item.title ?: "Boutique") },
                navigationIcon = {
                    IconButton(onClick = navigateUp) {
                        Icon(
                            imageVector = Icons.Filled.ArrowBack,
                            contentDescription = stringResource(R.string.app_back)
                        )
                    }
                }
            )
        }
        item {
            Spacer(modifier = Modifier.height(8.dp))
        }
        if (success != null) {
            item {
                Card(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(horizontal = 16.dp)
                        .padding(vertical = 4.dp),
                    colors = CardDefaults.cardColors(
                        containerColor = Color(0xFF0BDA51)
                    )
                ) {
                    Text(
                        text = success ?: "",
                        modifier = Modifier.padding(16.dp)
                    )
                }
            }
        }
        if (viewModel.item is TicketConfiguration && viewModel.item.userLeft != null && viewModel.item.userLeft!! < 1) {
            item {
                Card(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(horizontal = 16.dp)
                        .padding(vertical = 4.dp),
                    colors = CardDefaults.cardColors(
                        containerColor = MaterialTheme.colorScheme.primary
                    )
                ) {
                    Text(
                        text = "Cet élément n'est plus disponible.",
                        modifier = Modifier.padding(16.dp)
                    )
                }
            }
        }
        item {
            ShopCard(
                item = viewModel.item,
                detailsEnabled = false,
                cotisant = oldRootViewModel.getUser().value?.cotisant != null,
            )
        }
        item {
            Card(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 16.dp)
                    .padding(vertical = 4.dp)
            ) {
                Column(
                    modifier = Modifier
                        .padding(16.dp)
                        .fillMaxWidth()
                ) {
                    Text(
                        text = "Acheter",
                        fontWeight = FontWeight.Bold
                    )
                    ShopItemPrice(viewModel.item, user?.cotisant != null)
                    if (viewModel.item is TicketConfiguration && viewModel.item.userLeft != null) {
                        Text(
                            text = "${viewModel.item.userLeft} place(s) restante(s)",
                            modifier = Modifier.padding(vertical = 8.dp)
                        )
                    }
                    if (viewModel.item.canPayLater) {
                        Picker(
                            modifier = Modifier
                                .padding(vertical = 8.dp)
                                .fillMaxWidth(),
                            placeholder = "Méthode de paiement...",
                            items = mapOf(
                                true to "Lydia",
                                false to "A un membre BDE"
                            ),
                            selected = payNow ?: true,
                            onSelected = viewModel::setPayNow,
                        )
                    }
                    Button(
                        modifier = Modifier
                            .fillMaxWidth(),
                        enabled = loading != true,
                        onClick = {
                            viewModel.launchBuy(oldRootViewModel.getToken().value)
                        }
                    ) {
                        Text(text = "Acheter")
                    }
                    Text(
                        "Astuce : ajoutez votre adresse mail uha à votre compte Lydia pour simplifier le processus de paiement."
                    )
                    if (error != null) {
                        AlertDialog(
                            onDismissRequest = viewModel::dismissError,
                            title = { Text("Une erreur est survenue !") },
                            text = { Text(error ?: "Erreur inconnue") },
                            confirmButton = {
                                Button(onClick = viewModel::dismissError) {
                                    Text("OK")
                                }
                            }
                        )
                    }
                }
            }
        }
    }

}

File: ios/SuiteBDE/Features/Shop/ShopItemView.swift
//
//  ShopItemView.swift
//  BDE
//
//  Created by Nathan Fallet on 27/03/2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import SwiftUI
import shared

struct ShopItemView: View {
    
    @EnvironmentObject var rootViewModel: OldRootViewModel
    @StateObject var viewModel: ShopItemViewModel
    
    var body: some View {
        ScrollView {
            VStack(alignment: .leading) {
                if let success = viewModel.success {
                    Text(success)
                        .foregroundColor(.white)
                        .frame(maxWidth: .infinity)
                        .cardView(background: Color.green)
                }
                if let configuration = viewModel.item as? TicketConfiguration, Int64(truncating: configuration.userLeft ?? 1) < 1 {
                    Text("Cet élément n'est plus disponible.")
                        .foregroundColor(.white)
                        .frame(maxWidth: .infinity)
                        .cardView(background: Color.accentColor)
                }
                ShopCard(
                    item: viewModel.item,
                    detailsEnabled: false,
                    cotisant: rootViewModel.user?.cotisant != nil
                )
                VStack(alignment: .leading, spacing: 8) {
                    HStack {
                        Text("Acheter")
                            .fontWeight(.bold)
                        Spacer()
                    }
                    ShopItemPrice(item: viewModel.item, cotisant: rootViewModel.user?.cotisant != nil)
                    if let configuration = viewModel.item as? TicketConfiguration, let userLeft = configuration.userLeft {
                        Text("\(userLeft) place(s) restante(s)")
                    }
                    if viewModel.item.canPayLater {
                        Picker(selection: $viewModel.payNow, label: Text("")) {
                            Text("Lydia").tag(true)
                            Text("A un membre BDE").tag(false)
                        }
                        .pickerStyle(.segmented)
                    }
                    Button("Acheter") {
                        viewModel.launchBuy(token: rootViewModel.token)
                    }
                    .disabled(viewModel.loading)
                    .buttonStyle(FilledButtonStyle())
                    Text("Astuce : ajoutez votre adresse mail uha à votre compte Lydia pour simplifier le processus de paiement.")
                }
                .cardView()
            }
            .padding()
        }
        .navigationTitle(Text(viewModel.item.title ?? "Boutique"))
        .onAppear(perform: viewModel.onAppear)
        .sheet(item: $viewModel.url) { url in
            SafariView(url: URL(string: url)!)
        }
        .alert(item: $viewModel.error) { error in
            Alert(
                title: Text("Une erreur est survenue !"),
                message: Text(error)
            )
        }
    }
    
}
